File: codes/go/009_operator_optional_chaining/009_operator_optional_chaining.go
package main

import (
	"encoding/json"
	"fmt"
	"reflect"
)

const EMPTY_STRING = ""
const TAB = "    "

// type any interface{}
type array []any
type object map[string]any

func prettyJsonStringify(anything any) string {
    marshalledJson, err := json.MarshalIndent(anything, EMPTY_STRING, TAB)
    if (err == nil) {
        return string(marshalledJson)
    }
    return "undefined"
}

func arrayReduce(callbackFunction func(any, any, int, array) any, anArray array, initialValue any) any {
    // JavaScript-like Array.reduce() function
    result := initialValue
    for arrayItemIndex, arrayItem := range anArray {
        result = callbackFunction(result, arrayItem, arrayItemIndex, anArray)
    }
    return result
}

func optionalChaining(anything any, objectPropertiesArray ...any) any {
    anythingType := reflect.TypeOf(anything).Kind()
    if (((anythingType != reflect.Map) && (anythingType != reflect.Slice)) || (len(objectPropertiesArray) == 0)) {
        return anything
    }
    return arrayReduce(func(currentResult any, currentItem any, _ int, _ array) any {
        var currentItemTypeString any = nil
        switch currentItemType := reflect.TypeOf(currentItem).Kind(); currentItemType {
        case reflect.String:
            currentItemTypeString = "String"
        case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64, reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Float32, reflect.Float64, reflect.Complex64, reflect.Complex128:
            currentItemTypeString = "Number"
        default:
            currentItemTypeString = nil
        }
        if (currentItemTypeString == nil) {
            return nil
        }
        if (currentResult == nil && (anythingType == reflect.Map) && (currentItemTypeString == "String")) {
            return anything.(object)[currentItem.(string)]
        }
        if (currentResult == nil && (anythingType == reflect.Slice) && (currentItemTypeString == "Number") && (currentItem.(int) >= 0) && (len(anything.(array)) > currentItem.(int))) {
            return anything.(array)[currentItem.(int)]
        }
        currentResultType := reflect.TypeOf(currentResult).Kind()
        if (currentResultType == reflect.Map && (currentItemTypeString == "String")) {
            return currentResult.(object)[currentItem.(string)]
        }
        if ((currentResultType == reflect.Slice) && (currentItemTypeString == "Number") && (currentItem.(int) >= 0) && (len(currentResult.(array)) > currentItem.(int))) {
            return currentResult.(array)[currentItem.(int)]
        }
        return nil
    }, objectPropertiesArray, nil)
}

func main() {
    fmt.Println("\n// JavaScript-like Optional Chaining Operator (?.) in Go")

    // There's no JavaScript-like Optional Chaining Operator (?.) in Go.
    // But, we can create our own function to mimic it in Go.

    JSON_OBJECT := object{
        "foo": object{
            "bar": "baz",
        },
        "fruits": array{"apple", "mango", "banana"},
    }
    fmt.Println("JSON_OBJECT:", prettyJsonStringify(JSON_OBJECT))

    fmt.Println("JSON_OBJECT?.foo?.bar or JSON_OBJECT?.['foo']?.['bar']:", prettyJsonStringify(optionalChaining(JSON_OBJECT, "foo", "bar")))
    // JSON_OBJECT?.foo?.bar or JSON_OBJECT?.['foo']?.['bar']: "baz"

    fmt.Println("JSON_OBJECT?.foo?.baz or JSON_OBJECT?.['foo']?.['baz']:", prettyJsonStringify(optionalChaining(JSON_OBJECT, "foo", "baz")))
    // JSON_OBJECT?.foo?.baz or JSON_OBJECT?.['foo']?.['baz']: null

    fmt.Println("JSON_OBJECT?.fruits?.[2] or JSON_OBJECT?.['fruits']?.[2]:", prettyJsonStringify(optionalChaining(JSON_OBJECT, "fruits", 2)))
    // JSON_OBJECT?.fruits?.[2] or JSON_OBJECT?.['fruits']?.[2]: "banana"

    fmt.Println("JSON_OBJECT?.fruits?.[5] or JSON_OBJECT?.['fruits']?.[5]:", prettyJsonStringify(optionalChaining(JSON_OBJECT, "fruits", 5)))
    // JSON_OBJECT?.fruits?.[5] or JSON_OBJECT?.['fruits']?.[5]: null
}

File: codes/visual_basic_dotnet/009_operator_optional_chaining.vb
Imports System.Collections.Generic ' Dictionary, KeyValuePair

Module Program
    Function JsonStringify(ByVal Anything As Object, Optional ByVal Pretty As Boolean = False, Optional ByVal Indent As String = "    ") As String
        Dim IndentLevel As Integer = 0
        Dim JsonStringifyInner As Func(Of Object, String, String) = Function(ByVal AnythingInner As Object, ByVal IndentInner As String)
            If (AnythingInner Is Nothing) Then Return "null"
            If (TypeOf AnythingInner Is String) Then Return """" & AnythingInner & """"
            If (IsNumeric(AnythingInner) = True) Then Return CStr(AnythingInner).Replace(",", ".")
            If (TypeOf AnythingInner Is Boolean) Then Return CStr(AnythingInner)
            If (TypeOf AnythingInner Is List(Of Object)) Then
                If (AnythingInner.Count = 0) Then Return "[]"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "[" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "[")
                Dim ArrayItemIndex As Integer = 0
                For Each ArrayItem As Object In AnythingInner
                    Result &= JsonStringifyInner(ArrayItem, IndentInner)
                    If ((ArrayItemIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ArrayItemIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "]", "]")
                Return Result
            End If
            If (TypeOf AnythingInner Is Dictionary(Of String, Object)) Then
                If (AnythingInner.Count = 0) Then Return "{}"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "{" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "{")
                Dim ObjectIterationIndex As Integer = 0
                For Each ObjectEntry As KeyValuePair(Of String, Object) In AnythingInner
                    Dim ObjectKey = ObjectEntry.Key
                    Dim ObjectValue = ObjectEntry.Value
                    Result &= """" & ObjectKey & """: " & JsonStringifyInner(ObjectValue, IndentInner)
                    If ((ObjectIterationIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ObjectIterationIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "}", "}")
                Return Result
            End If
            Return "null"
        End Function
        Return JsonStringifyInner(Anything, Indent)
    End Function

    Function OptionalChaining(ByVal Anything As Object, ByVal ParamArray ObjectPropertiesArray() As Object) As Object
        ' JavaScript-like Optional Chaining Operator (?.) function
        If ((((TypeOf Anything Is Dictionary(Of String, Object)) = false) AndAlso ((TypeOf Anything Is List(Of Object)) = false)) OrElse (ObjectPropertiesArray.Length = 0)) Then Return Anything
        Dim ArrayReduceCallback As Func(Of Object, Object, Object) = Function(ByVal CurrentResult As Object, ByVal CurrentItem As Object) As Object
            If ((CurrentResult Is Nothing) AndAlso (TypeOf Anything Is Dictionary(Of String, Object)) AndAlso (TypeOf CurrentItem Is String) AndAlso DirectCast(Anything, Dictionary(Of String, Object)).ContainsKey(CStr(CurrentItem))) Then
                Try
                    Return DirectCast(Anything, Dictionary(Of String, Object))(CStr(CurrentItem))
                Catch Ex As Exception
                    Return Nothing
                End Try
            End If
            If ((CurrentResult Is Nothing) AndAlso (TypeOf Anything Is List(Of Object)) AndAlso (IsNumeric(CurrentItem) = true) AndAlso (CInt(CurrentItem) >= 0) AndAlso (DirectCast(Anything, List(Of Object)).Count > CInt(CurrentItem))) Then
                Try
                    Return DirectCast(Anything, List(Of Object))(CInt(CurrentItem))
                Catch Ex As Exception
                    Return Nothing
                End Try
            End If
            If ((TypeOf CurrentResult Is Dictionary(Of String, Object)) AndAlso (TypeOf CurrentItem Is String) AndAlso DirectCast(CurrentResult, Dictionary(Of String, Object)).ContainsKey(CStr(CurrentItem))) Then
                Try
                    Return DirectCast(CurrentResult, Dictionary(Of String, Object))(CStr(CurrentItem))
                Catch Ex As Exception
                    Return Nothing
                End Try
            End If
            If ((TypeOf CurrentResult Is List(Of Object)) AndAlso (IsNumeric(CurrentItem) = true) AndAlso (CInt(CurrentItem) >= 0) AndAlso (DirectCast(CurrentResult, List(Of Object)).Count > CInt(CurrentItem))) Then
                Try
                    Return DirectCast(CurrentResult, List(Of Object))(CInt(CurrentItem))
                Catch Ex As Exception
                    Return Nothing
                End Try
            End If
            Return Nothing
        End Function
        Return ObjectPropertiesArray.Aggregate(Nothing, ArrayReduceCallback)
    End Function

    Sub Main(Args As String())
        Console.WriteLine($"{Environment.NewLine}' JavaScript-like Optional Chaining Operator (?.) in Visual Basic (.NET)")

        ' There's no JavaScript-like Optional Chaining Operator (?.) in Visual Basic (.NET).
        ' But, we can create our own function to mimic it in Visual Basic (.NET).

        Dim JSON_OBJECT As Object = New Dictionary(Of String, Object) From {
            {"foo", New Dictionary(Of String, Object) From {
                {"bar", "baz"}
            }},
            {"fruits", New List(Of Object) From {"apple", "mango", "banana"}}
        }
        Console.WriteLine("JSON_OBJECT: " & JsonStringify(JSON_OBJECT, Pretty:=True))

        Console.WriteLine("' using JavaScript-like Optional Chaining Operator (?.) function ""OptionalChaining""")

        Console.WriteLine($"JSON_OBJECT?.foo?.bar or JSON_OBJECT?.['foo']?.['bar']: {JsonStringify(OptionalChaining(JSON_OBJECT, "foo", "bar"))}")
        ' JSON_OBJECT?.foo?.bar or JSON_OBJECT?.['foo']?.['bar']: baz

        Console.WriteLine($"JSON_OBJECT?.foo?.baz or JSON_OBJECT?.['foo']?.['baz']: {JsonStringify(OptionalChaining(JSON_OBJECT, "foo", "baz"))}")
        ' JSON_OBJECT?.foo?.baz or JSON_OBJECT?.['foo']?.['baz']: null

        Console.WriteLine($"JSON_OBJECT?.fruits?.[2] or JSON_OBJECT?.['fruits']?.[2]: {JsonStringify(OptionalChaining(JSON_OBJECT, "fruits", 2))}")
        ' JSON_OBJECT?.fruits?.[2] or JSON_OBJECT?.['fruits']?.[2]: "banana"

        Console.WriteLine($"JSON_OBJECT?.fruits?.[5] or JSON_OBJECT?.['fruits']?.[5]: {JsonStringify(OptionalChaining(JSON_OBJECT, "fruits", 5))}")
        ' JSON_OBJECT?.fruits?.[5] or JSON_OBJECT?.['fruits']?.[5]: null
    End Sub
End Module
